presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPath.java

```diff
@@ -13,19 +13,146 @@
 
+import com.facebook.presto.common.function.JsonPathExtractionEngine;
+import com.facebook.presto.spi.PrestoException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.InvalidJsonException;
+import com.jayway.jsonpath.InvalidPathException;
+import com.jayway.jsonpath.PathNotFoundException;
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider;
 import io.airlift.slice.Slice;
 
+import java.io.IOException;
+import java.io.InputStream;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.airlift.slice.Slices.utf8Slice;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
 public class JsonPath
 {
     private final JsonExtract.JsonExtractor<Slice> scalarExtractor;
     private final JsonExtract.JsonExtractor<Slice> objectExtractor;
     private final JsonExtract.JsonExtractor<Long> sizeExtractor;
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static JsonExtract.JsonExtractor<Slice> getScalarExtractorForJayway(com.jayway.jsonpath.JsonPath jsonPath, Configuration jaywayConfig)
+    {
+        return new JsonExtract.JsonExtractor<Slice>()
+        {
+            @Override
+            public Slice extract(InputStream inputStream)
+                    throws IOException
+            {
+                JsonNode node = jaywayExtract(jsonPath, jaywayConfig, inputStream);
+                if (node == null || !node.isValueNode()) {
+                    return null;
+                }
+                return utf8Slice(node.asText());
+            }
+        };
+    }
+
+    private static JsonExtract.JsonExtractor<Slice> getObjectExtractorForJayway(com.jayway.jsonpath.JsonPath jsonPath, Configuration jaywayConfig)
+    {
+        return new JsonExtract.JsonExtractor<Slice>()
+        {
+            @Override
+            public Slice extract(InputStream inputStream)
+                    throws IOException
+            {
+                JsonNode node = jaywayExtract(jsonPath, jaywayConfig, inputStream);
+                if (node == null) {
+                    return null;
+                }
+                return utf8Slice(node.toString());
+            }
+        };
+    }
+
+    private static JsonExtract.JsonExtractor<Long> getSizeExtractorForJayway(com.jayway.jsonpath.JsonPath jsonPath, Configuration jaywayConfig)
+    {
+        return new JsonExtract.JsonExtractor<Long>()
+        {
+            @Override
+            public Long extract(InputStream inputStream)
+                    throws IOException
+            {
+                JsonNode node = jaywayExtract(jsonPath, jaywayConfig, inputStream);
+                if (node == null) {
+                    return null;
+                }
+                return (long) node.size(); // Jackson correctly returns 0 for scalar nodes
```

Comment:
What does the comment indicate here ? Does jayway has a different behavior than the existing Json?

Reply:
The presto doc indicates that json_size should return 0 for scalar values. I meant this is also what the Jackson JsonNode does so we don't need any special if case here.